### ISSUE TITLE
Audible difference in music playback

### ISSUE THREAD

**Issue by Nindaleth** [Try/OpenGothic#117]
Lastly I've been spending some hours in Valley of Mines in the original and have gotten used to the one song playing in most of the area. When reaching the area in OpenGothic, the music sounds different, although the same data should be used for playback as in the original.

For the comparison, I've recorded two videos, in both I've loaded a save standing in a similar place, so that the music starts to play from the beginning. It can be heard throughout the video that OpenGothic sounds differently for some reason, but the most audible part is between 1:00 and 1:15. Another one is the drum sounding a lot quieter between 0:00 and 0:10.

Notes:
* Both original and OpenGothic will play the menu music for a while on startup. After it finishes, I reload the save to force a fresh start of the music track. OpenGothic restarts the music ~0.5s after the loading is finished, that's why the previous state of the loop can be heard shortly at the beginning - I don't mind that and don't consider that an issue at the moment.
* I've noticed that readme lists the Music as in progress, is this issue the reason or is there something else?

Original:
https://user-images.githubusercontent.com/6383928/106178749-71fa0280-619a-11eb-962e-e3aa3a5c1b23.mp4

OpenGothic:
https://user-images.githubusercontent.com/6383928/106178725-6c042180-619a-11eb-9433-7041176caa9b.mp4

**Comment by Try** [comment]
> OpenGothic sounds differently for some reason

There is more than "some" reason :D Most likely there is more than one issue in my implementation of DirectX music.

Original uses DXMusic from DirectX7, OpenGothic - emulate DX8 music as good as it can.
DX8 format structure is based on [DLS-sound](https://en.wikipedia.org/wiki/DLS_format) format + some stuff on top of it.

To keep it short: 
* Music, consists of PCM/ADPCM sound samples - those are simple and done reliably.
* Sound samples grouped by bands(`Dx8::Band`)
* Instrument(`Dx8::PatternList::Instrument`) = list of notes + curves + reference to band
* Pattern = array of instruments

OpenGothic handles pattern switch by itself and instrument rendering by library TinySoundFont.

The soundtrack :
![изображение](https://user-images.githubusercontent.com/122894/106191156-c9a06a00-61aa-11eb-84ba-0819e8a4900f.png)

The drummer represented by instrument №22 and №23.

The issue:
Sound samples of drummer plays same way in DirectMusicProducer and OpenGothic, but the whole instrument has big difference in sound volume: OpenGothic is very quiet. I don't see a reason why instrument 22, 23 should be more laugh - I'm probably missing something big. 

Note: I'm taking DirectMusicProducer is a reference implementation, as it probably most 'correct' one.
Note2: all sources related to music can be found in game/dmusic.
Note3: it's been more than a year since I've done most of current music implementation - already forgot some things.
Note4: OpenGothic renders instrument fist and apples curves later - this is probably wrong.

**Comment by Nindaleth** [comment]
Whoa, that's an explanation I didn't expect! I thought Gothic was mixing together a few MP3s or something, not that there is full-blown DX8-provided music tracker.

But your explanation does seem to point to "only" one reason - everything I can hear wrong with the current state could be explained by improper volume leveling. There are no missing (or invalid) instruments or incorrect durations - the drums are just too quiet and the strings are just too loud. On further listening, I can hear now that the violin sound plays for that long time even in the original, the difference is that in original the volume curve goes downward and the violin thus gets quieter in the last few seconds.

**Comment by VitalyPetrov** [comment]
@Try hi!
Do you have any ideas how to fix this issue? I tried the last build and it's the only significant issue as for me :(

**Comment by Try** [comment]
Hi, @VitalyPetrov !

No good ideas unfortunately. Ideally we would like to use third party library for music, yet there seem to be no such for DX-music.
Can you also point so particular soundtracks to check/test, other than mine-valley?

PS:
"debugging" by listening is horrible experience :D

**Comment by hwnde** [comment]
@Try Perhaps DX music implementation from Wine could help? 
https://gitlab.winehq.org/wine/wine/-/tree/master/dlls/dmusic?ref_type=heads

**Comment by JaXt0r** [comment]
Hi Try. Did you investigate [libdmusic](https://github.com/libdmusic/libdmusic)? It was built in 2019 and mentioned to be the music implementation for [ReGoth](https://github.com/REGoth-project/REGoth). If feasible, it should be easy to use for OpenGothic as kind of ReGoth's successor. 😀

**Comment by Try** [comment]
Hi, @JaXt0r !

Yes, been there, done that. libdmusic has same level of completeness as ReGoth. For Hall of Idorath music was turn into bit-noise.

**Comment by DramaDesu** [comment]
Hello there!
Are you still interesting in fixing that?

I'm trying to find out what's going on, but got some results.
It's not correct yet, but much better that it was.

https://github.com/user-attachments/assets/57af22dd-7b94-4aab-89c2-9516ad5dc154

**Comment by DramaDesu** [comment]
https://github.com/user-attachments/assets/48f2e320-f996-4898-b3bc-bc45c261e50f

**Comment by Try** [comment]
Hi, @DramaDesu !

> Are you still interesting in fixing that?

Yes, while work with music is on background track improvements are always welcome. Can you please clarify what is it on your videos? Is it some UE5 plugin?

**Comment by DramaDesu** [comment]
Hello, @Try !
Yes, it's but it's only for UI and testing, I've moved dx music backend.
As soon as I get a satisfactory result, I’ll merge it with the repo and add a test via the console. There’s already a test, but no autocomplete, which makes it a bit of a hassle to search for musical themes.
Most tracks sounds great for now, but KAS_DayStd, trying to understand how to fix mix, if play plain, or render only pattern it works.

**Comment by DramaDesu** [comment]
Trumpets sound great yet, but it works
[https://www.youtube.com/watch?v=oW7KAXpVM80](url)

_Could anyone tell me how to insert any link properly?_

**Comment by DramaDesu** [comment]
Main issues are sdl, mapping for instruments (like in direct music, they're really tricky), envelopes, lfo, pitch
I hear clipping will fix that
Not great results, but seems much better, the original sound is the best, I've tried to play it from VM on Direct Music Producer, it's really helpful, glanced at decompiling switch version, it seems to me like they've fixed dmusic lib or something like that, but didn't make transitions, they works extremely great in the game, I wish they could do that to find out how it works, but the company who worked on the port just did simply fadeout

The original is still much better, significantly

**Comment by Try** [comment]
Sound really great!

Is it meant to be a brand new sound backend or is it path to existing(GothicKit) one?

> Could anyone tell me how to insert any link properly?

Leaving link as plain text usually fine

**Comment by bugproof** [comment]
https://github.com/tongzx/nt5src/tree/master/Source/XPSP1/NT/multimedia/directx/dmusic/

There seems to be a code of original dmusic available? Maybe just throw it at good LLM and let it figure out possible issue? https://github.com/tongzx/nt5src/blob/master/Source/XPSP1/NT/multimedia/directx/dmusic/dmusic/dminstru.cpp

**Comment by DramaDesu** [comment]
> Sound really great!
> 
> Is it meant to be a brand new sound backend or is it path to existing(GothicKit) one?
> 
> > Could anyone tell me how to insert any link properly?
> 
> Leaving link as plain text usually fine

Thanks, but still some effects don't work properly
It's the current backend, not the GothicKit one

**Comment by DramaDesu** [comment]
> https://github.com/tongzx/nt5src/tree/master/Source/XPSP1/NT/multimedia/directx/dmusic/
> 
> There seems to be a code of original dmusic available? Maybe just throw it at good LLM and let it figure out possible issue? https://github.com/tongzx/nt5src/blob/master/Source/XPSP1/NT/multimedia/directx/dmusic/dmusic/dminstru.cpp

What a treasure, I'll try that

**Comment by DramaDesu** [comment]
@Try @bugproof 
I've tried, it works, we could call it "dmusic slop", but the result is quite good.
Please, check this out: [https://www.youtube.com/watch?v=thREgCJEMro](url)

**Comment by DramaDesu** [comment]
It seems to me, like all effects work, include LFO, and pitch bending
I'm not sure about portability, cause it has been moved from the sources, but will handle it
LLM made the whole port, 35 new files has been added, and most of internal dmusic lib stuff have been changed
I wish I could have another OS, but I'll try to compile the project on WSL

**Comment by Try** [comment]
> I wish I could have another OS, but I'll try to compile the project on WSL

You can just push PR - that will trigger automatic build in CI, for x86, arm, win64/linux/mac.

**Comment by DramaDesu** [comment]
It compiles and work on wsl
Okay, I'll do a draft, I think the code could smell

**Comment by bugproof** [comment]
> It compiles and work on wsl Okay, I'll do a draft, I think the code could smell

Good job, the code could smell but imo LLM models at current stage do a much better job that humans would do for those kind of tedious porting tasks - so I would argue that human code smelt more otherwise we wouldn't have that issue. The original code is readily available. Also if you use opus with 1M context. Now it just takes a couple more LLM passes to ensure it ported everything bit by bit but as you can see from the video it sounds much better than any human attemps re-implementing dmusic 100% by hand over the last couple of years. If dmusic had some tests that would make it easier to check.

One thing to note tho. The code contains "Copyright Microsoft" so not sure if it won't be a problem here - can we consider dmusic abandonware?

**Comment by DramaDesu** [comment]
Yes, totally agree, LLM for porting work perfect. I've rechecked the port with opus. 
It might be worth adding some tests to check that the track parsing is working properly, so we can see which features are being used and which aren’t; it might not work on other versions of dmusic, but it works for Gothic.

With "Copyright Microsoft" is a kind of risk, might better do a private repo, but public compiled libs.
I think I was a bit hasty.
Anyway it needs to be fixed for macos.

**Comment by DramaDesu** [comment]
Another option is simply to rewrite it
But as you said, it might smell more

**Comment by Try** [comment]
Hi, @DramaDesu ! I've took a small look at pr (62 files!).  

The nt5 folder - looks like a it a rip off from somewhere? Unfortunately direct copy of else code wont be merged. 
Also by the look of PR, it's not in direct use (at least I seen no includes). Is it some sort of reference? 

Also, to note: @lmichaelis been working on music solution(GothicKit) for OpenGothic in the past. In your PR, you introducing dxsynth, and it's already done in GothicKit. Can't quite reason about completeness state of both, but both seem to work.
Basically raising awareness, for you of one-another, to avoid duplication of work. And, I think, contributing to gothickit can be more future proof, since my old OG music driver is basically bunch of hacks.

**Comment by DramaDesu** [comment]
Hi, @Try!
Good point, I think, I should have done a draft pr, just for testing.
Speaking of nt5, it's a port for original direct music.
I think, I will contribute on GothicKit, it's more reasonable, than direct copy.
We’ll need to figure out exactly what the problem is, but from what I’ve gathered, it’s the instrument mappings and various dynamic effects—they have quite a significant impact—plus the built-in reverb.

Anyway, I'll try focusing on GothicKit than.

**Comment by bugproof** [comment]
@DramaDesu  Just an opinion but I think it would be more beneficial to create a separate cross-platform dmusic library since it's not strictly Gothic related but many old games might benefit from this too. Licensing-wise it's hard to question, technically re-implementation and reverse engineering could be illegal too. Same goes for OpenGothic. For example Take-Two Interactive shutdown GTA reimplementation project - we don't really know what THQ Nordic is capable of. You could as well ask Opus to re-implement it in a way thats different, just preserve the essential algorithms - you can't really do 100% different because then the sound will be different too and that's not the goal. Since the source is already leaked it would be stupid not to use it. [Wine project](https://www.winehq.org/) could benefit from this too.

Edit: oh i think GothicKit also has it as separate lib https://github.com/GothicKit/dmusic I wonder if @lmichaelis is aware the source code of dmusic exists or if he's doing actual reversing from binaries work/guessing

**Comment by lmichaelis** [comment]
Using source code to re-implemented proprietary libraries is very risky even if the software itself is no longer distributed and Microsoft is very litigious @bugproof. That's why neither the Wine-team nor I want to use it in any re-implementation.

**Comment by DramaDesu** [comment]
I'd been "fixing" the first improvements (like mappings, envelopes, reverb) without any source code, only OpenGothic and DirectMusic Producer, I loaded Gothic sequences to the Producer, opened instruments to check sound, played segment to verify, I managed better sound but it wasn't enough,  for now I have no idea, how to handle it, since the source code has leaked, might do something like a clean room, but it's a huge project.

**Comment by bugproof** [comment]
> Using source code to re-implemented proprietary libraries is very risky even if the software itself is no longer distributed and Microsoft is very litigious [@bugproof](https://github.com/bugproof). That's why neither the Wine-team nor I want to use it in any re-implementation.

Could argue that re-implementation by itself is very risky because technically we don't have any rights neither for Gothic nor dmusic - never forget what happened to GTA 3+VC re-implementation. THQ Nordic has also a right at any point to DMCA this repo even if it's not original copy of Gothic source code. At this point I doubt Microsoft even cares. dmusic can be considered abandonware and using leaks to solve 5 year old issues is probably the best way forward. Hell - Microsoft doesn't even care about pirated copies of Windows - project like MASSGRAVE exists for years.

**Comment by Spiffyk** [comment]
> never forget what happened to GTA 3+VC re-implementation

That GTA re-implementation was shut down precisely because they used the original binaries, decompiled, to achieve what they did, essentially making their engine a derivative work of the original one. If you look at OpenRW, a proper reverse-engineered engine for GTA, just like OpenGothic, that one is fine (even though not super active right now).
